src/main/java/com/facebook/openwifirrm/ucentral/UCentralClient.java

```diff
@@ -195,27 +199,142 @@ public boolean login() {
 		}
 
 		// Parse access token from response
-		JSONObject respBody;
+		WebTokenResult token;
 		try {
-			respBody = new JSONObject(response.getBody());
-		} catch (JSONException e) {
+			token = gson.fromJson(response.getBody(), WebTokenResult.class);
+		} catch (JsonSyntaxException e) {
 			logger.error("Login failed: Unexpected response", e);
 			logger.debug("Response body: {}", response.getBody());
 			return false;
 		}
-		if (!respBody.has("access_token")) {
+		if (
+			token == null || token.access_token == null ||
+				token.access_token.isEmpty()
+		) {
 			logger.error("Login failed: Missing access token");
-			logger.debug("Response body: {}", respBody.toString());
+			logger.debug("Response body: {}", response.getBody());
 			return false;
 		}
-		this.accessToken = respBody.getString("access_token");
+		this.accessToken = token;
 		logger.info("Login successful as user: {}", username);
-		logger.debug("Access token: {}", accessToken);
+		logger.debug("Access token: {}", accessToken.access_token);
+		logger.debug("Refresh token: {}", accessToken.refresh_token);
 
 		// Load system endpoints
 		return loadSystemEndpoints();
 	}
 
+	/**
+	 * when using public endpoints, refresh the access token if it's expired.
+	 */
+	public synchronized void refreshAccessToken() {
+		if (usePublicEndpoints) {
+			refreshAccessTokenImpl();
+		}
+	}
+
+	/**
+	 * Check if the token is completely expired even if
+	 * for a token refresh request
+	 *
+	 * @return true if the refresh token is expired
+	 */
+	private boolean isAccessTokenExpired() {
+		if (accessToken == null) {
+			return true;
+		}
+		return accessToken.created + accessToken.expires_in <
+			Instant.now().getEpochSecond();
+	}
+
+	/**
+	 * Check if an access token is expired.
+	 *
+	 * @return true if the access token is expired
+	 */
+	private boolean isAccessTokenTimedOut() {
+		if (accessToken == null) {
+			return true;
+		}
+		return accessToken.created + accessToken.idle_timeout <
```

Comment:
I think we need to keep track of last accessed time and then use `last_accessed + accessToken.idle_timeout` since idle timeout only starts applying from last access